daffodil-core/src/main/scala/org/apache/daffodil/core/dpath/Expression.scala

```diff
@@ -744,21 +742,38 @@ abstract class PathExpression() extends Expression {
    * about an array such as its current count.
    */
   lazy val isPathToOneWholeArray: Boolean = {
-    if (steps == Nil) false // root is never an array
-    if (steps == Nil) false // root is never an array
-    else if (isSelf) false // self path is never an array
-    else
-      steps.last.isArray &&
-      !steps.last.pred.isDefined && // last cannot have a [N] pred
-      steps.dropRight(1).forall {
-        // for all the prior steps
-        // if they mention an array element, there
-        // must be a [N] predicate.
-        step =>
-          if (step.isInstanceOf[Up]) true
-          else if (step.isArray) step.pred.isDefined
-          else true
-      }
+    isNotRootOrSelf &&
+    steps.last.isArray &&
+    priorStepsHaveIndexingPredicates
+  }
+
+  /**
+   * Path to a single array element without a [N] predicate qualifying it
+   * or to an optional element.
+   *
+   * That is, the kind of path expression used to access information
+   * about an array or optional such as its current count.
+   */
+  override lazy val isPathToOneWholeArrayOrOptional: Boolean = {
+    isNotRootOrSelf &&
+    steps.last.isArrayOrOptional &&
+    priorStepsHaveIndexingPredicates
+  }
+
+  private lazy val isNotRootOrSelf: Boolean = {
+    if (steps == Nil) false
+    else if (isSelf) false
+    else true
+  }
+
+  private lazy val priorStepsHaveIndexingPredicates: Boolean = {
```

Comment:
Can this val be removed? With the `isLastStep` addition in `NamedStep`, I think we should now SDE if any prior steps are arrays without an index predicate, so now the `steps.dropRight(1).forAll` check should always evaluate to true an isn't really needed. I think we do still need the `steps.last.pred.isEmpty` check but that probably makes more sense as an added check in `isPathToOneWholeArrayOrOptional` and `isPathToOneWholeArray`.

Reply:
This priorStepsHaveIndexingPredicates expression is required in order to properly process array elements. With this removed, we saw several tests fail.

Reply:
Did you add `steps.last.pred.isEmpty` to `isPathToOneWholeArrayOrOptional` and `isPathToOneWholeArray`?

Reply:
Good catch!  After this update, the priorStepsHaveIndexingPredicates expression is no longer needed.